OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,115 @@ public void Tick(Squad owner)
-				if (target != null)
+				if (TryMakeWay > 0)
 				{
-					owner.TargetActor = target;
-					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+					owner.Bot.QueueOrder(new Order("AttackMove", pathGuider, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+					foreach (var a in owner.Units)
+					{
+						if (a != pathGuider)
+							owner.Bot.QueueOrder(new Order("Scatter", a, false));
+					}
+
+					TryMakeWay--;
 				}
 				else
-					foreach (var a in owner.Units)
-						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					// When going through is over, restore the check
+					StuckInPath = StuckInPathCheckTimes + MakeWayTick;
+					TryMakeWay = MakeWayTick;
+				}
+
+				return;
 			}
 
-			if (ShouldFlee(owner))
-				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
+			// 3. Check if the squad is stucked due to the map has very twisted path
+			// or currently bridge and tunnel from TS mod
+			if ((pathGuider.CenterPosition - LastPos).LengthSquared <= 4)
+				StuckInPath--;
+			else
+				StuckInPath = StuckInPathCheckTimes;
+
+			LastPos = pathGuider.CenterPosition;
+
+			// 4. Since units have different movement speeds, they get separated while approaching the target.
+
+			/* Let them regroup into tighter formation towards "pathGuider".
+			 *
+			 * "unitsArea" means the space the squad units will occupy (if 1 per Cell).
+			 * pathGuider only stop when scope of "unitsAround" is not covered all units;
+			 * units in "unitsHurryUp"  will catch up,
+			 * which keep the team tight while not stucked.
+			 *
+			 * Imagining "unitsArea" takes up a a place shape like square, we need to draw a circle
+			 * to cover the the enitire circle.
+			 *
+			 * When look around, pathGuider find units around to decide if it need to continue.
+			 * and units that need hurry up will try catch up before guider waiting
+			 *
+			 * However in practice because of the poor PF, squad tend to PF to a eclipse.
+			 * "lookAround" now has the radius of two times of the circle mentioned before.
+			 */
+
+			var groupArea = (long)WDist.FromCells(owner.Units.Count).Length * 1024;
+
+			var unitsHurryUp = owner.Units.Where(a => (a.CenterPosition - pathGuider.CenterPosition).LengthSquared >= groupArea * 2).ToArray();
+			var lookAround = owner.Units.Where(a => (a.CenterPosition - pathGuider.CenterPosition).LengthSquared <= groupArea * 5).ToArray();
+
+			if (owner.Units.Count > lookAround.Length)
+				owner.Bot.QueueOrder(new Order("Stop", pathGuider, false));
+			else
+				owner.Bot.QueueOrder(new Order("AttackMove", pathGuider, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+			foreach (var unit in unitsHurryUp)
+				owner.Bot.QueueOrder(new Order("AttackMove", unit, Target.FromCell(owner.World, pathGuider.Location), false));
```

Comment:
This could be made more fluid by giving a guard order instead of an attackMove order. The behaviour would be the same except they will keep following the current position of the pathGuide instead of the position where the pathGuide was when the order was given

Reply:
I used to use guard here,  but there can be some actors are unguardable in other mod(without guardable trait)  and cause game crash.  use AttackMove will be a relatively secured idea.

Reply:
Other mods including CA, SP.

Reply:
You could check if the `pathGuide` has the `Guardable` trait first and use AttackMove as a backup.

Reply:
Ummm.. I'd rather to make the code simple here, since now AttackMove is now first Attack then Move, and although not very fluid, AttackMove is enough to keep a relatively tight formation without blocking guide way if there are too many units (can be solved by stuck check but I would rather not to let it happen, Guard order will let units follow closely and can block way if >20 units for a squad, even in regular human gameplay).

Reply:
P.s
You can change parameter in  `unitsHurryUp` to `>= 1` which will be like what guard order will do. In this case you may find if slow unit are selected as guide, fast units will always block its way.